adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/providers/ChirpStackProvider.java

```diff
@@ -0,0 +1,152 @@
+ *******************************************************************************/
+
+package org.eclipse.hono.adapter.lora.providers;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.hono.adapter.lora.LoraConstants;
+import org.eclipse.hono.adapter.lora.LoraMessageType;
+import org.springframework.stereotype.Component;
+
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A LoRaWAN provider with API for ChirpStack.
+ */
+@Component
+public class ChirpStackProvider implements LoraProvider {
+
+    private static final String FIELD_CHIRPSTACK_PAYLOAD = "data";
+    private static final String FIELD_CHIRPSTACK_DEVICE = "devEUI";
+    private static final String FIELD_CHIRPSTACK_TX_INFO = "txInfo";
+    private static final String FIELD_CHIRPSTACK_SPREADING_FACTOR = "spreadingFactor";
+    private static final String FIELD_CHIRPSTACK_BANDWIDTH = "bandwidth";
+    private static final String FIELD_CHIRPSTACK_FUNCTION_PORT = "fPort";
+    private static final String FIELD_CHIRPSTACK_CODE_RATE = "codeRate";
+    private static final String FIELD_CHIRPSTACK_LORA_MODULATION_INFO = "loRaModulationInfo";
+    private static final String FIELD_CHIRPSTACK_FREQUENCY = "frequency";
+    private static final String FIELD_CHIRPSTACK_FRAME_COUNT = "fCnt";
+    private static final String FIELD_CHIRPSTACK_RX_INFO = "rxInfo";
+    private static final String FIELD_CHIRPSTACK_GATEWAY_ID = "gatewayID";
+    private static final String FIELD_CHIRPSTACK_RSSI = "rssi";
+    private static final String FIELD_CHIRPSTACK_LSNR = "loRaSNR";
+    private static final String FIELD_CHIRPSTACK_CHANNEL = "channel";
+    private static final String FIELD_CHIRPSTACK_LOCATION = "location";
+    private static final String FIELD_CHIRPSTACK_LATITUDE = "latitude";
+    private static final String FIELD_CHIRPSTACK_LONGITUDE = "longitude";
+
+    @Override
+    public String getProviderName() {
+        return "chirpStack";
+    }
+
+    @Override
+    public String pathPrefix() {
+        return "/chirpstack";
+    }
+
+    @Override
+    public String extractDeviceId(final JsonObject loraMessage) {
+        return loraMessage.getString(FIELD_CHIRPSTACK_DEVICE);
+    }
+
+    @Override
+    public String extractPayload(final JsonObject loraMessage) {
+        return loraMessage.getString(FIELD_CHIRPSTACK_PAYLOAD);
+    }
+
+    @Override
+    public LoraMessageType extractMessageType(final JsonObject loraMessage) {
+        if (loraMessage.containsKey(FIELD_CHIRPSTACK_PAYLOAD)) {
+            return LoraMessageType.UPLINK;
+        }
+        return LoraMessageType.UNKNOWN;
+    }
+
+    @Override
+    public Map<String, Object> extractNormalizedData(final JsonObject loraMessage) {
+        final Map<String, Object> returnMap = new HashMap<>();
+        if (loraMessage.containsKey(FIELD_CHIRPSTACK_TX_INFO)) {
+            final JsonObject txInfo = loraMessage.getJsonObject(FIELD_CHIRPSTACK_TX_INFO);
```

Comment:
I do not think that you need to look up the key twice ...
the same is true for all of the other properties you are retrieving in the remainder of the method.